modules/images/webp-uploads/load.php

```diff
@@ -683,3 +683,121 @@ function webp_uploads_update_rest_attachment( WP_REST_Response $response, WP_Pos
+
+		if ( ! file_exists( $image_file ) ) {
+			continue;
+		}
+
+		$metadata['sources'][ $targeted_mime ] = array(
+			'file'     => wp_basename( $image_file ),
+			'filesize' => filesize( $image_file ),
+		);
+
+		foreach ( $metadata['sizes'] as $size => $size_details ) {
+			// Add sources to resized image metadata.
+			$filename_without_extension = pathinfo( $size_details['file'], PATHINFO_FILENAME );
+			$image_file                 = path_join( $original_directory, "$filename_without_extension.{$extension[0]}" );
+
+			if ( ! file_exists( $image_file ) ) {
+				continue;
+			}
+
+			$metadata['sizes'][ $size ]['sources'][ $targeted_mime ] = array(
+				'file'     => wp_basename( $image_file ),
+				'filesize' => filesize( $image_file ),
+			);
+		}
+	}
+
+	return $metadata;
+}
+
+/**
+ *
+ * Creates additional image formats when original image is edited.
+ *
+ * @since n.e.x.t
+ *
+ * @param bool|null       $override  Value to return instead of saving. Default null.
+ * @param string          $file      Name of the file to be saved.
+ * @param WP_Image_Editor $image     The image editor instance.
+ * @param string          $mime_type The mime type of the image.
+ * @param int             $post_id   Attachment post ID.
+ *
+ * @return bool|null
+ */
+function webp_uploads_update_image_onchange( $override, $file, $image, $mime_type, $post_id ) {
+	if ( null !== $override ) {
+		return $override;
+	}
+
+	if ( empty( webp_uploads_get_upload_image_mime_transforms()[ $mime_type ] ) ) {
+		return $override;
+	}
+
+	$valid_mime_transforms = webp_uploads_get_upload_image_mime_transforms()[ $mime_type ];
+	$original_directory    = pathinfo( $file, PATHINFO_DIRNAME );
+	$filename              = pathinfo( $file, PATHINFO_FILENAME );
+	$sizes                 = wp_get_registered_image_subsizes();
+	$allowed_mimes         = array_flip( wp_get_mime_types() );
+
+	foreach ( $valid_mime_transforms as $targeted_mime ) {
+		if ( $targeted_mime === $mime_type ) {
+			continue;
+		}
+
+		if ( ! isset( $allowed_mimes[ $targeted_mime ] ) || ! is_string( $allowed_mimes[ $targeted_mime ] ) ) {
+			return $override;
+		}
+
+		if ( ! wp_image_editor_supports( array( 'mime_type' => $targeted_mime ) ) ) {
+			return $override;
+		}
+
+		$extension   = explode( '|', $allowed_mimes[ $targeted_mime ] );
+		$destination = trailingslashit( $original_directory ) . "{$filename}.{$extension[0]}";
+
+		if ( is_wp_error( $image->save( $destination, $targeted_mime ) ) ) {
+			return $override;
+		}
+
+		$_sizes = array();
+
+		foreach ( $sizes as $size => $size_details ) {
+			$_sizes[ $size ] = $size_details;
+		}
+
+		$image->multi_resize( $_sizes );
+	}
+	add_filter(
+		'wp_update_attachment_metadata',
+		function ( $metadata, $post_meta_id ) use ( $post_id, $valid_mime_transforms, $allowed_mimes, $file ) {
+			if ( $post_meta_id !== $post_id ) {
+				return $metadata;
+			}
+
+			return webp_uploads_update_sources( $metadata, $valid_mime_transforms, $allowed_mimes, $file );
```

Comment:
See my comment above, it would be more appropriate to generate the additional image sub-sizes right here, before calling `webp_uploads_update_sources()`. We need to only generate sub-sizes for those in `$metadata['sizes']` that differ from those in `wp_get_attachment_metadata( $post_meta_id )['sizes']`.

Moving the logic itself should be fairly trivial, we can simply pass the `$image` variable (the `WP_Image_Editor` instance) through to the callback here as well.

Also, per my other comment, we should pass the additional full image's new filename to this callback so that we can use it to populate the main `$metadata['sources']`.

I think at that point I'm not sure how useful `webp_uploads_update_sources()` to just update the sources still is. Again, the primary purpose of that function right now is to "guess" the new image file names, but it would be much more reliable to directly use the return value of the `WP_Image_Editor::multi_resize()` call.